src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java

```diff
@@ -0,0 +1,187 @@
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static com.netgrif.application.engine.startup.ImpersonationRunner.IMPERSONATION_CONFIG_PETRI_NET_IDENTIFIER;
+
+@Service
+public class ImpersonationAuthorizationService implements IImpersonationAuthorizationService {
+
+    public static final int MAX_CONFIGS_PER_USER = 1000;
+
+    @Autowired
+    private IUserService userService;
+
+    @Autowired
+    private IElasticCaseService elasticCaseService;
+
+    @Autowired
+    private IAuthorityService authorityService;
+
+    @Autowired
+    private IWorkflowService workflowService;
+
+    @Autowired
+    private IProcessRoleService processRoleService;
+
+    @Override
+    public Page<IUser> getConfiguredImpersonationUsers(String query, LoggedUser impersonator, Pageable pageable) {
+        if (impersonator.isAdmin()) {
+            return userService.searchAllCoMembers(query, null, null, impersonator, true, pageable);
+
+        } else {
+            Page<Case> cases = searchConfigs(impersonator.getId(), pageable);
+            List<IUser> users = cases.getContent().stream()
+                    .map(c -> ((UserFieldValue) c.getDataSet().get("impersonated").getValue()).getId())
+                    .distinct()
+                    .map(id -> userService.findById(id, true))
+                    .collect(Collectors.toList());
+            return new PageImpl<>(users, pageable, cases.getTotalElements());
+        }
+    }
+
+    @Override
+    public boolean canImpersonate(LoggedUser impersonator, String configId) {
+        Case config = getConfig(configId);
+        return isValidAndContainsUser(config, impersonator.getId());
+    }
+
+    @Override
+    public boolean canImpersonateUser(LoggedUser impersonator, String userId) {
+        IUser impersonated = userService.findById(userId, true);
+        List<Case> impersonationConfig = searchConfigs(impersonator.getId(), impersonated.getStringId());
+        return impersonator.isAdmin() || !impersonationConfig.isEmpty();
+    }
+
+    @Override
+    public Page<Case> searchConfigs(String impersonatorId, Pageable pageable) {
+        return findCases(request(impersonatorId, null), pageable);
+    }
+
+    @Override
+    public List<Case> searchConfigs(String impersonatorId, String impersonatedId) {
+        Page<Case> cases = findCases(request(impersonatorId, impersonatedId), PageRequest.of(0, MAX_CONFIGS_PER_USER));
+        return cases.getContent();
+    }
+
+    @Override
+    public List<Authority> getAuthorities(List<Case> configs, IUser impersonated) {
+        if (configs.isEmpty()) {
+            return new ArrayList<>();
+        }
+        List<String> authIds = (List) configs.get(0).getDataSet().get("impersonated_authorities").getValue();
```

Comment:
what about the authorities from the other config elements?

Reply:
this is an oversight. This code branch is not actually used atm (frontend does not utilize this in this specific context) but I fixed it nonetheless